src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private T[] arr;
+
+    public ArrayList() {
+        arr = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size < arr.length) {
+            arr[size] = value;
+            size++;
+        } else {
+            resize();
+            arr[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        if (index == arr.length || size >= arr.length) {
+            resize();
+        }
 
+        T[] newArr = (T[]) new Object[arr.length];
+        System.arraycopy(arr, 0, newArr, 0, size);
+        newArr[index] = value;
+        System.arraycopy(arr, index, newArr, index + 1, size - index);
+        System.arraycopy(newArr, 0, arr, 0, newArr.length);
+        if (index >= 0 && index <= size) {
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        return arr[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        arr[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T[] newArr;
+        T result = null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        result = arr[index];
+        newArr = (T[]) new Object[arr.length];
+        System.arraycopy(arr, 0, newArr, 0, index);
+        System.arraycopy(arr, index + 1, newArr, index, arr.length - index - 1);
+        System.arraycopy(newArr, 0, arr, 0, arr.length);
+        size--;
+        return result;
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(T element) throws NoSuchElementException {
+        boolean found = false;
+
+        for (int i = 0; i < size; i++) {
+            if (arr[i] == element || (arr[i] != null && arr[i].equals(element))) {
+                remove(i);
+                return element;
```

Comment:
Using '==' for comparison is not safe for non-primitive types. Use 'equals' method for object comparison. Also, consider using the extracted method for element removal by index to avoid code duplication.